lib/spack/spack/test/llnl/util/tty/log.py

```diff
@@ -0,0 +1,472 @@
+    for i in range(10):
+        _mock_shell_integration_no_termios(proc, ctl, attrs)
+    os.kill(proc.pid, signal.SIGUSR1)
+
+
+@contextlib.contextmanager
+def no_termios():
+    saved = llnl.util.tty.log.termios
+    llnl.util.tty.log.termios = None
+    yield
+    llnl.util.tty.log.termios = saved
+
+
+@contextlib.contextmanager
+def nullcontext():
+    yield
+
+
+@pytest.mark.skipif(not which("ps"), reason="requires ps utility")
+@pytest.mark.skipif(not termios, reason="requires termios support")
+@pytest.mark.parametrize('test_fn,termios_on_or_off', [
+    # tests with termios
+    (mock_shell_fg, nullcontext),
+    (mock_shell_bg, nullcontext),
+    (mock_shell_bg_fg, nullcontext),
+    (mock_shell_fg_bg, nullcontext),
+    (mock_shell_tstp_cont, nullcontext),
+    (mock_shell_tstp_tstp_cont, nullcontext),
+    (mock_shell_tstp_tstp_cont_cont, nullcontext),
+    # tests without termios
+    (mock_shell_fg_no_termios, no_termios),
+    (mock_shell_bg, no_termios),
+    (mock_shell_bg_fg_no_termios, no_termios),
+    (mock_shell_fg_bg_no_termios, no_termios),
+    (mock_shell_tstp_cont, no_termios),
+    (mock_shell_tstp_tstp_cont, no_termios),
+    (mock_shell_tstp_tstp_cont_cont, no_termios),
+])
+def test_foreground_background_simple(test_fn, termios_on_or_off):
+    shell = PseudoShell(test_fn, mock_logger)
+    shell.attrs["debug"] = True
+
+    # run the shell test
+    with termios_on_or_off():
+        shell.start()
+    exitcode = shell.join()
+
+    # processes completed successfully
+    assert exitcode == 0
+
+
+@pytest.mark.skipif(not which("ps"), reason="requires ps utility")
+@pytest.mark.skipif(not termios, reason="requires termios support")
+@pytest.mark.parametrize('test_fn,termios_on_or_off', [
+    # tests with termios
+    (mock_shell_v_v, nullcontext),
+    (mock_shell_integration_1, nullcontext),
+    (mock_shell_integration_loop, nullcontext),
+    # tests without termios
+    (mock_shell_v_v_no_termios, no_termios),
+    (mock_shell_integration_1_no_termios, no_termios),
+    (mock_shell_integration_loop_no_termios, no_termios),
+])
+def test_foreground_background_with_output(test_fn, capfd, termios_on_or_off):
+    shell = PseudoShell(test_fn, mock_logger)
+    shell.attrs["debug"] = True
+
+    with termios_on_or_off():
+        shell.start()
+    exitcode = shell.join()
+
+    out, err = capfd.readouterr()
+    print("stderr ===========================")
+    print(err)
+    print()
+    print("stdout ===========================")
+    print(out)
+    print()
+
+    # processes completed successfully
+    assert exitcode == 0
+
+    # numbers output by logger
+    stripped = out.strip()
+    split = re.split(r'\s+', stripped) if stripped else []
+
+    numbers = set([int(n) for n in split])
+    all_numbers = set(range(shell.attrs["count"]))
+
+    # test that logger.force_echo() works
+    assert 0 in numbers
+
+    # all numbers are really output we expect
+    assert numbers <= all_numbers
```

Comment:
I think I see that this test operates by constantly outputting new numbers while switching terminal output, but this approach feels a bit random: would it be possible to use interprocess signals to have the child process note when it reads a specific number, and we could ensure that exactly the numbers we expect are present?

Reply:
This is hard to set up because `v` is fundamentally asynchronous w.r.t. the output of the logged process.  I think the test as-is is pretty simple.  I do not see a way to make it more specific as you suggest without making it very invasive.

Reply:
As it is, I think there are many possible ways this test can fail without it being obvious in the test output, so my goal is to figure out how to make this more precise.

I agree my initial suggestion is lacking: I've had trouble in the past when I tried coordinating with interprocess signals to check output. The problem is that it is hard to reason about whether the signals are interfering or creating an artificial result.

I think it might be easier to improve this test with locking: we can create a lock that is held by the parent process whenever it is in "v" mode. The child process checks this lock and if it is held, then it outputs "2", and if it is not, then it outputs "1". In more detail:

* There would actually have to be two locks, one for when "v" is applied and one for when the child process is going to write something, let's call that "w"
* The parent process must hold "w" before it acquires or releases "v"
* The child process must hold "w" before it writes anything

Does that seem more reasonable?
I think that would be easier to reason about than IPC signals.

Reply:
Can you elaborate on the non-obvious ways this could fail?  Essentially this types v, sleeps, and types v again, and it looks to see that something was output. It actually needs a check to make sure more than just 0 was output, but I am not seeing how your approach helps. It would require me to modify log_output, at which point I’m not really testing the “real” code anymore. I don’t think it’s worth it.  I also don’t see why it needs to be “precise”. The feature is not precise. We are testing whether there is output or not, not what that output is, and that is fine here.

Reply:
I'm not talking about how this test can fail: I'm saying that there are many things that could go wrong in the source itself, and this test would not catch it, or it would fail and it would be difficult to understand what went wrong.

> It would require me to modify log_output

I don't think either of my suggested approaches require that. I'm talking about coordinating between the entity that sends the "v" signal, and the child process that is doing the printing.

> I also don’t see why it needs to be “precise”

If anything goes wrong, and the test is not precise, then it is difficult to figure out how to fix it: anyone editing this code or fixing bugs in the future will have trouble understanding how to fix the test.

> We are testing whether there is output or not, not what that output is

There are many ways we can end up with no output. Most of them are bugs. We want to end up with no output in exactly the way we desire. Being precise about what is captured will help us understand whether we actually succeeded.

Reply:
@scheibelp: the tests have been simplified and use locks as you suggest, but I do not think this is so much better than before.  In particular, locks don't get rid of the races here.  It is still possible to have a race where the child process notices `v` on `master_fd` *after* the lock takes effect.

I implemented this so that the child writes `off` when `v` is not enabled and `on` when it is -- you can still get some `off`'s in the output, and I do not see a non-invasive way around that, nor do I really think we should try hard to do that.

The new version does make no output unambiguously wrong.  So that's good.  But I had to put a kludgy condition on the assertions -- I allow up to 2 `off`'s in the output.  I think this is about the best we can do and would rather not spend a lot more time on it.  The important piece of this test is that it verifies that `v` toggles output on and off.  I think the current tests do that well enough.